### PR TITLE
suggest doubling recursion limit in more situations

### LINKED ISSUE #38852: Rustc does not recommend raising the recursion limit
It used to be that when you reach the macro recursion limit, rustc would recommend raising it and give the attribute to do so. This has been lost.

Expected Result:

Diagnostic tells how to increase recursion limit.

Actual Result:

No note/help.

Reproduce [(playground gist)](https://play.rust-lang.org/?gist=450998d6abe14d9d90e131ef9af79796&version=nightly&backtrace=0)

```rust
macro_rules! recurse {
    () => { };
    ($t:tt $($tail:tt)*) => { recurse!($($tail)*) };
}

fn main() {
    recurse!(0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9
             0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9
             0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9
             0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9);
}
```

### PR DESCRIPTION
**Pull request by durka**
Fixes #38852.

r? @bluss

### CODE REVIEW

File: src/test/ui/did_you_mean/recursion_limit_deref.stderr [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  --> $DIR/recursion_limit_deref.rs:56:22
+   |
+56 |     let x: &Bottom = &t;
+   |                      ^^ deref recursion limit reached
+   |
+   = help: consider adding a `#[recursion_limit="20"]` attribute to your crate
+
+error[E0055]: reached the recursion limit while auto-dereferencing I
+  |
+  = help: consider adding a `#[recursion_limit="20"]` attribute to your crate
```
Comment (durka) [inline src/test/ui/did_you_mean/recursion_limit_deref.stderr]: This duplicate error shouldn't be here and yet it is. cc #38940 @jseyfried

**Comment by durka** [comment]
r?

File: src/libsyntax/ext/base.rs [hunk 2]
```diff
@@ -674,9 +674,15 @@ impl<'a> ExtCtxt<'a> {
...
-            self.span_fatal(ei.call_site,
-                            &format!("recursion limit reached while expanding the macro `{}`",
-                                    ei.callee.name()));
+            let suggested_limit = self.ecfg.recursion_limit * 2;
+            let mut err = self.struct_span_fatal(ei.call_site,
+                &format!("recursion limit reached while expanding the macro `{}`",
+                         ei.callee.name()));
+            err.note(&format!(
+                "consider adding a `#![recursion_limit=\"{}\"]` attribute to your crate",
+                suggested_limit));
```
Comment (bluss) [inline src/libsyntax/ext/base.rs]: Can this be `.help()` too or is there a reason it can't?

Comment (durka) [inline src/libsyntax/ext/base.rs]: Yes, I think it can be.

File: src/test/ui/did_you_mean/recursion_limit.stderr [hunk 3]
```diff
@@ -0,0 +1,21 @@
+error[E0275]: overflow evaluating the requirement `K: std::marker::Send`
+  --> $DIR/recursion_limit.rs:44:5
+   |
+44 |     is_send::<A>();
+   |     ^^^^^^^^^^^^
+   |
+   = note: consider adding a `#![recursion_limit="20"]` attribute to your crate
```
Comment (bluss) [inline src/test/ui/did_you_mean/recursion_limit.stderr]: This will need update to `help` I think

**Comment by bluss** [comment]
Thanks. I'll approve this after checking with rustc devs

File: src/libsyntax/ext/base.rs [hunk 4]
```diff
@@ -674,9 +674,15 @@ impl<'a> ExtCtxt<'a> {
...
-            self.span_fatal(ei.call_site,
-                            &format!("recursion limit reached while expanding the macro `{}`",
-                                    ei.callee.name()));
+            let suggested_limit = self.ecfg.recursion_limit * 2;
+            let mut err = self.struct_span_fatal(ei.call_site,
+                &format!("recursion limit reached while expanding the macro `{}`",
+                         ei.callee.name()));
+            err.help(&format!(
+                "consider adding a `#![recursion_limit=\"{}\"]` attribute to your crate",
+                suggested_limit));
```
Comment (nikomatsakis) [inline src/libsyntax/ext/base.rs]: maybe we should factor this code into some sort of helper?

Comment (nikomatsakis) [inline src/libsyntax/ext/base.rs]: I feel like it'd also be nice to make this into an error code ("recursion limit reached") with a note giving the specific details. That way, we could give an extended error description.

Comment (durka) [inline src/libsyntax/ext/base.rs]: Where should I put the helper and diagnostic code? Note that the PR is split across libsyntax and librustc_typeck.

Comment (nikomatsakis) [inline src/libsyntax/ext/base.rs]: Hmm. Can we only put diagnostic codes into rustc crates? I'm not sure.

Comment (durka) [inline src/libsyntax/ext/base.rs]: There are diagnostic codes in libsyntax. Can I define it there and use the same one from librustc_typeck?

Comment (nikomatsakis) [inline src/libsyntax/ext/base.rs]: Meh. I think we should improve this, but it's not that big a deal.

**Comment by durka** [comment]
Should be fixed.

**Pull request merged**